### PR TITLE
optimizing ADD; POP and similar

### PR DESCRIPTION
**Pull request by pirapira**
This pull-request adds a peep-hole optimization that replaces `ADD; POP` with `POP; POP`.  This enables some further optimizations.

Example:
```
contract C {
   function f (int a) {
       a + a;
   }
}
```

Before this PR, without `--optimize` option, the compiler produces `...DUP1 DUP2 ADD POP...`

After this PR, without `--optimize` option, this sequence disappears.

### CODE REVIEW

File: libevmasm/PeepholeOptimiser.cpp [hunk 1]
```diff
@@ -57,6 +57,31 @@ struct PushPop
...
+	static bool apply(AssemblyItems::const_iterator _in, std::back_insert_iterator<AssemblyItems> _out)
+	{
+		if (_in[1] == Instruction::POP &&
+			_in[0].type() == Operation
+		)
+		{
+			Instruction i0 = _in[0].instruction();
+			if (instructionInfo(i0).ret == 1 &&
+				instructionInfo(i0).args == 2 &&
+				!SemanticInformation::invalidatesMemory(i0) &&
```
Comment (chriseth) [inline libevmasm/PeepholeOptimiser.cpp:71]: Please also check for side-effects.

File: libevmasm/PeepholeOptimiser.cpp [hunk 2]
```diff
@@ -57,6 +57,31 @@ struct PushPop
...
+	static size_t windowSize() { return 2; }
+	static bool apply(AssemblyItems::const_iterator _in, std::back_insert_iterator<AssemblyItems> _out)
+	{
+		if (_in[1] == Instruction::POP &&
+			_in[0].type() == Operation
+		)
+		{
+			Instruction i0 = _in[0].instruction();
+			if (instructionInfo(i0).ret == 1 &&
+				instructionInfo(i0).args == 2 &&
```
Comment (chriseth) [inline libevmasm/PeepholeOptimiser.cpp]: We don't have to limit this to 2 arguments - we might even combine this with `PushPop`.

**Review by chriseth — CHANGES_REQUESTED** [review]
This case should actually be handled by the CommonSubexpressionEliminator, but sometimes, it cannot produce shorter code or cannot produce code at all, so it is good to have this.

File: libevmasm/PeepholeOptimiser.cpp (hunk 2 above)
Comment (pirapira) [reply libevmasm/PeepholeOptimiser.cpp]: Generalized this into arbitrary number of arguments.

File: libevmasm/PeepholeOptimiser.cpp (hunk 1 above)
Comment (pirapira) [reply libevmasm/PeepholeOptimiser.cpp:71]: I looked around and found the sideeffects flag.

**Review by chriseth — APPROVED** [review]

**Pull request merged**